File: FacebookInterviewQuestions/lruCache.js
class Node {
  constructor() {
    this.key = null;
    this.value = null;
    this.prev = null;
    this.next = null;
  }
}

class LRUCache {
  constructor(capacity) {
    this.capacity = capacity;

    this.head = new Node();
    this.tail = new Node();
    this.head.next = this.tail;
    this.tail.prev = this.head;

    this.cache = new Map();
    this.size = 0;
  }

  get(key) {
    if (this.cache.has(key)) {
      let node = this.cache.get(key);

      this.moveToHead(node);

      return node.value;
    }

    return -1;
  }

  put(key, value) {
    if (this.cache.has(key)) {
      const node = this.cache.get(key);
      node.value = value;
      this.moveToHead(node);
    } else {
      const node = new Node();
      node.key = key;
      node.value = value;

      this.cache.set(key, node);
      this.addToFront(node);
      this.size++;

      if (this.size > this.capacity) {
        this.removeLRUEntry();
      }
    }
  }

  removeLRUEntry() {
    const tail = this.popTail();
    this.cache.delete(tail.key);
    this.size--;
  }

  popTail() {
    const tail = this.tail.prev;
    this.removeNode(tail);
    return tail;
  }

  removeNode(node) {
    const prev = node.prev;
    const next = node.next;

    prev.next = node.next;
    next.prev = prev;
  }

  addToFront(node) {
    node.prev = this.head;
    node.next = this.head.next;

    this.head.next.prev = node;
    this.head.next = node;
  }

  moveToHead(node) {
    this.removeNode(node);
    this.addToFront(node);
  }
}

const cache = new LRUCache(2);

cache.put(1, 1);
cache.put(2, 2);
cache.get(1); // returns 1
cache.put(3, 3); // evicts key 2
cache.get(2); // returns -1 (not found)
cache.put(4, 4); // evicts key 1
cache.get(1); // returns -1 (not found)
cache.get(3); // returns 3
cache.get(4); // returns 4

File: BloombergInterviewQuestions/lruCache.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None
        
        
# we use a Doubly Linked List and a Hashmap
# DLL allows us to remove and put in front in O(1)
# hashmap allows us do a get in O(1)

class LRUCache:

    def __init__(self, capacity: int):
        self.size = 0
        self.capacity = capacity
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.cache = {}
        

    def get(self, key: int) -> int:
        node = self.cache.get(key)
        if node:
            # move node to head. 1 Detach Node, 2. Move to front
            self.deleteNode(node)
            self.addToFront(node)
            return node.val
        return -1
        

    def put(self, key: int, value: int) -> None:
        if key not in self.cache:
            node = Node(key, value)
            self.cache[key] = node
            self.addToFront(node)
            self.size += 1
            
            if self.size > self.capacity:
                
                # remove tail
                lastNode = self.tail.prev
                self.deleteNode(lastNode)
                del self.cache[lastNode.key]
        else:
            node = self.cache.get(key)
            node.val = value
            self.deleteNode(node)
            self.addToFront(node)
            
                
        
        
    def deleteNode(self, node):
        prev = node.prev
        
        nextNode = node.next
        prev.next = nextNode
        nextNode.prev = prev
        
    def addToFront(self, node):
        nextNode = self.head.next
        self.head.next = node
        node.prev = self.head
        node.next = nextNode
        nextNode.prev = node
        
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)